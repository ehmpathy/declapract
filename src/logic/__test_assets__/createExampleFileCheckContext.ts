import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
