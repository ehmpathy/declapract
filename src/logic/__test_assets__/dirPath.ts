export const testAssetsDirectoryPath = __dirname;
