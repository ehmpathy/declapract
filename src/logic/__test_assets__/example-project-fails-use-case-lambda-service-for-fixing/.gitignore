# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

