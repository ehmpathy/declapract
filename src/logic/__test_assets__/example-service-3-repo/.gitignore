.declapract
node_modules
package-lock.json
