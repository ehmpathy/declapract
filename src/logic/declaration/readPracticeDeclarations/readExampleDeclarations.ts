import { ExampleDeclaration } from '../../../domain/objects/ExampleDeclaration';
import { listPathsInDirectory } from '../../../utils/filepaths/listPathsInDirectory';

export const readExampleDeclarations = async ({
  declarationsRootDirectory,
  declaredExamplesDirectory,
}: {
  declarationsRootDirectory: string;
  declaredExamplesDirectory: string;
}) => {
  const exampleDirectories = await listPathsInDirectory({
    directory: declaredExamplesDirectory,
  });
  const examples = exampleDirectories.map((directory) => {
    const name = directory;
    return new ExampleDeclaration({
      name,
      exampleRootDirectory: `${declaredExamplesDirectory}/${directory}`.replace(
        `${declarationsRootDirectory}/`,
        '',
      ),
    });
  });
  return examples;
};
