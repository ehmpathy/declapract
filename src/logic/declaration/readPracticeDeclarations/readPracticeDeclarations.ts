import { listPathsInDirectory } from '../../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../../UserInputError';
import { readPracticeDeclaration } from './readPracticeDeclaration/readPracticeDeclaration';

export const readPracticeDeclarations = async ({
  declaredPracticesDirectory,
}: {
  declaredPracticesDirectory: string;
}) => {
  const practiceDirectories = await listPathsInDirectory({
    directory: declaredPracticesDirectory,
  });
  if (!practiceDirectories.length)
    throw new UserInputError(
      `at least one practice needs to be defined in the practices directory: '${declaredPracticesDirectory}'`,
    );
  const practices = await Promise.all(
    practiceDirectories.map((directory) =>
      readPracticeDeclaration({
        declaredPracticeDirectory: `${declaredPracticesDirectory}/${directory}`,
      }),
    ),
  );
  return practices;
};
