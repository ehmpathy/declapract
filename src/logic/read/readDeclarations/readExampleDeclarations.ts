import { ExampleDeclaration } from '../../../domain/objects/ExampleDeclaration';
import { listPathsInDirectory } from '../../../utils/filepaths/listPathsInDirectory';

export const readExampleDeclarations = async ({ declaredExamplesDirectory }: { declaredExamplesDirectory: string }) => {
  const exampleDirectories = await listPathsInDirectory(declaredExamplesDirectory);
  const examples = exampleDirectories.map((directory) => {
    const name = directory;
    return new ExampleDeclaration({
      name,
    });
  });
  return examples;
};
