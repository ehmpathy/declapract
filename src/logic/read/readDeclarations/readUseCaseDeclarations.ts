import uniqBy from 'lodash.uniqby';
import { PracticeDeclaration, UseCaseDeclaration } from '../../../domain';
import { ExampleDeclaration } from '../../../domain/objects/ExampleDeclaration';
import { UseCasesDeclarationInput } from '../../../domain/objects/UseCasesDeclarationInput';
import { readYmlFile } from '../../../utils/fileio/readYmlFile';
import { UserInputError } from '../../UserInputError';

export const readUseCaseDeclarations = async ({
  declaredUseCasesPath,
  practices,
}: {
  declaredUseCasesPath: string;
  practices: PracticeDeclaration[];
}) => {
  // try to read the file
  const useCasesYml = await readYmlFile({ filePath: declaredUseCasesPath });

  // read and validate the useCases yml file
  const useCasesInput = await (async () => {
    try {
      return new UseCasesDeclarationInput(useCasesYml);
    } catch (error) {
      throw new UserInputError('use-cases declaration file contents failed validation', {
        potentialSolution: error.message,
      });
    }
  })();

  // now build up the use cases
  const useCases = Object.entries(useCasesInput['use-cases']).map(
    ([name, definition]) =>
      new UseCaseDeclaration({
        name,
        practices: definition.practices.map((requestedPracticeName) => {
          const foundPractice = practices.find((practice) => practice.name === requestedPracticeName);
          if (!foundPractice)
            throw new UserInputError(
              `A use-case specified a practices that was not defined. 'use-case:${name}' specified 'practice:${requestedPracticeName}', but no practice with name '${requestedPracticeName}' has been defined`,
              {
                potentialSolution: 'Have you checked that there is not a typo in the practice name?',
              },
            );
          return foundPractice;
        }),
        example: definition.example ? new ExampleDeclaration({ name: definition.example }) : null,
      }),
  );

  // now hydrate the use cases that requested "extends" of another use case
  const hydratedUseCases = useCases.map((useCase) => {
    // determine whether this use case asked for an extension
    const { extends: extendsUseCaseNames } = useCasesInput['use-cases'][useCase.name];
    if (!extendsUseCaseNames) return useCase; // if it doesn't extend any use cases, then we can stop here

    // for each use case it extends and build up the full, hydrated set of practices
    const hydratedPractices = [...useCase.practices];
    extendsUseCaseNames.forEach((extendsUseCaseName) => {
      // lookup the use case that it extends
      const extendedUseCase = useCases.find((candidateUseCase) => candidateUseCase.name === extendsUseCaseName);
      if (!extendedUseCase)
        throw new UserInputError(
          `A use-case was defined to extend a non-existent use-case. 'use-case:${useCase.name}' was defined to extend 'use-case:${extendsUseCaseName}', but no use case with name '${extendsUseCaseName}' is defined`,
          { potentialSolution: 'Was there a typo?' },
        );

      // push the practices on that extended use case
      hydratedPractices.push(...extendedUseCase.practices);
    });

    // now dedupe them by name
    const uniqueHydratedPractices = uniqBy(hydratedPractices, (practice) => practice.name);

    // and return the hydrated use case
    return new UseCaseDeclaration({ ...useCase, practices: uniqueHydratedPractices });
  });

  // and return the use cases
  return hydratedUseCases;
};
