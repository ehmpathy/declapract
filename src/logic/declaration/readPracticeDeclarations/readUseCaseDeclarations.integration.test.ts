import { testAssetsDirectoryPath } from '../../__test_assets__/dirPath';
import { readExampleDeclarations } from './readExampleDeclarations';
import { readPracticeDeclarations } from './readPracticeDeclarations';
import { readUseCaseDeclarations } from './readUseCaseDeclarations';

describe('readUseCaseDeclarations', () => {
  it('should get the declarations correctly', async () => {
    const practices = await readPracticeDeclarations({
      declaredPracticesDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices`,
    });
    const examples = await readExampleDeclarations({
      declaredExamplesDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/examples`,
    });
    const useCases = await readUseCaseDeclarations({
      declaredUseCasesPath: `${testAssetsDirectoryPath}/example-best-practices-repo/src/useCases.yml`,
      practices,
      examples,
    });
    expect(useCases.length).toEqual(2);
    expect(useCases).toMatchSnapshot();
  });
});
