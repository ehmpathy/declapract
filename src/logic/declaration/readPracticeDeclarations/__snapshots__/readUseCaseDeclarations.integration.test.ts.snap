// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readUseCaseDeclarations should get the declarations correctly 1`] = `
Array [
  UseCaseDeclaration {
    "example": null,
    "name": "typescript",
    "practices": Array [
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": ".prettierignore",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "package.json",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "prettier.config.js",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
          ],
          "name": "prettier",
          "readme": null,
        },
        "name": "prettier",
      },
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "src/**/sleep.ts",
              "purpose": "BEST_PRACTICE",
              "required": false,
              "type": "EQUALS",
            },
          ],
          "name": "util-sleep",
          "readme": "checks that if a file named \`sleep.ts\` exists, then it must match the proper definition for our sleep utility, defined as a best practice

optional check => if you dont need \`sleep.ts\`, dont define the file

helpful in that:
- it makes sure your definition of sleep is correct
- you can just define an empty file called \`sleep.ts\` and then run \`declapract fix --practice:util-sleep\` and it will fill it out for you
",
        },
        "name": "util-sleep",
      },
    ],
  },
  UseCaseDeclaration {
    "example": ExampleDeclaration {
      "exampleRootDirectory": "src/examples/lambda-service",
      "name": "lambda-service",
    },
    "name": "lambda-service",
    "practices": Array [
      PracticeDeclaration {
        "badPractices": Array [
          ProjectCheckDeclaration {
            "checks": Array [
              FileCheckDeclaration {
                "check": [Function],
                "fix": null,
                "pathGlob": "package.json",
                "purpose": "BAD_PRACTICE",
                "required": true,
                "type": "CUSTOM",
              },
            ],
            "name": "moment",
            "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
          },
        ],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": null,
              "pathGlob": "package.json",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CUSTOM",
            },
          ],
          "name": "dates-and-times",
          "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
        },
        "name": "dates-and-times",
      },
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": null,
              "pathGlob": "package.json",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CUSTOM",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "serverless.yml",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
          ],
          "name": "serverless",
          "readme": null,
        },
        "name": "serverless",
      },
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": ".gitignore",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": ".terraform-version",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "provision/terraform/product/variables.tf",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "provision/terraform/environments/dev/main.tf",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "provision/terraform/environments/dev/versions.tf",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "provision/terraform/environments/prod/main.tf",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "provision/terraform/environments/prod/versions.tf",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
          ],
          "name": "terraform",
          "readme": "best pattern found for terraform so far is to put different environments into their own directory - and just have them reference a shared product dir
",
        },
        "name": "terraform",
      },
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": ".prettierignore",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "package.json",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "CONTAINS",
            },
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "prettier.config.js",
              "purpose": "BEST_PRACTICE",
              "required": true,
              "type": "EQUALS",
            },
          ],
          "name": "prettier",
          "readme": null,
        },
        "name": "prettier",
      },
      PracticeDeclaration {
        "badPractices": Array [],
        "bestPractice": ProjectCheckDeclaration {
          "checks": Array [
            FileCheckDeclaration {
              "check": [Function],
              "fix": [Function],
              "pathGlob": "src/**/sleep.ts",
              "purpose": "BEST_PRACTICE",
              "required": false,
              "type": "EQUALS",
            },
          ],
          "name": "util-sleep",
          "readme": "checks that if a file named \`sleep.ts\` exists, then it must match the proper definition for our sleep utility, defined as a best practice

optional check => if you dont need \`sleep.ts\`, dont define the file

helpful in that:
- it makes sure your definition of sleep is correct
- you can just define an empty file called \`sleep.ts\` and then run \`declapract fix --practice:util-sleep\` and it will fill it out for you
",
        },
        "name": "util-sleep",
      },
    ],
  },
]
`;
